include/pybind11/embed.h

```diff
@@ -87,29 +89,103 @@ struct embedded_module {
     }
 };
 
+struct wide_char_arg_deleter {
+    void operator()(wchar_t* ptr) const {
+#if PY_VERSION_HEX >= 0x030500f0
+        // API docs: https://docs.python.org/3/c-api/sys.html#c.Py_DecodeLocale
+        PyMem_RawFree(ptr);
+#else
+        delete[] ptr;
+#endif
+    }
+};
+
+inline wchar_t* widen_chars(char* safe_arg) {
+#if PY_VERSION_HEX >= 0x030500f0
+    wchar_t* widened_arg = Py_DecodeLocale(safe_arg, nullptr);
+#else
+    wchar_t* widened_arg = nullptr;
+#  if HAVE_BROKEN_MBSTOWCS
+    size_t count = strlen(safe_arg);
+#  else
+    size_t count = mbstowcs(nullptr, safe_arg, 0);
+#  endif
+    if (count != static_cast<size_t>(-1)) {
+        widened_arg = new wchar_t[count + 1];
+        mbstowcs(widened_arg, safe_arg, count + 1);
+    }
+#endif
+    return widened_arg;
+}
+
+/// Python 2.x/3.x-compatible version of `PySys_SetArgv`
+inline void set_interpreter_argv(int argc, char** argv, bool add_current_dir_to_path) {
+    // Before it was special-cased in python 3.8, passing an empty or null argv
+    // caused a segfault, so we have to reimplement the special case ourselves.
+    char** safe_argv = argv;
+    std::unique_ptr<char*[]> argv_guard;
+    std::unique_ptr<char[]> argv_inner_guard;
+    if (nullptr == argv || argc <= 0) {
+        argv_guard = std::unique_ptr<char*[]>(safe_argv = new char*[1]);
+        argv_inner_guard = std::unique_ptr<char[]>(safe_argv[0] = new char[1]);
+        safe_argv[0][0] = '\0';
+        argc = 1;
+    }
+#if PY_MAJOR_VERSION >= 3
+    size_t argv_size = static_cast<size_t>(argc);
+    // SetArgv* on python 3 takes wchar_t, so we have to convert.
+    std::unique_ptr<wchar_t*[]> widened_argv(new wchar_t*[argv_size]);
+    std::vector< std::unique_ptr<wchar_t[], wide_char_arg_deleter> > widened_argv_entries;
+    for (size_t ii = 0; ii < argv_size; ++ii) {
+        widened_argv_entries.emplace_back(widen_chars(safe_argv[ii]));
+        if (!widened_argv_entries.back()) {
+            // A null here indicates a character-encoding failure or the python
+            // interpreter out of memory. Give up.
+            return;
```

Comment:
Should we raise an exception here, instead of silently doing nothing?

Reply:
I think silently doing nothing is actually OK here. Consider:
- All this means is that `PySys_SetArgvEx` won't be called, but current pybind doesn't call it, so probably 95% of users won't notice or care.
- Assuming locales are configured correctly on your machine, you are probably out of memory. As such, your exception handler is not likely to work.
- Assuming locales are configured _incorrectly_, a failure here will be harder to debug than the inevitable failures in the user's python code.

That said, I'm not married to the decision. It just seemed like a reasonable fallback to revert to the old behavior.